src/neo/Network/P2P/TaskManager.cs

```diff
@@ -274,56 +224,68 @@ public static Props Props(NeoSystem system)
+                    return;
                 }
-                if (!AssignSyncTask(failedTask)) return;
             }
 
-            int taskCounts = sessions.Values.Sum(p => p.IndexTasks.Count);
-            var highestBlockIndex = sessions.Values.Max(p => p.LastBlockIndex);
-            for (; taskCounts < MaxSyncTasksCount; taskCounts++)
+            // When the number of AvailableTasks is no more than 0, no pending tasks of InventoryType.Block, it should process pending the tasks of headers
+            // If not HeaderTask pending to be processed it should ask for more Blocks
+            if ((!HasHeaderTask || globalTasks[HeaderTaskHash] < MaxConncurrentTasks) && Blockchain.Singleton.HeaderHeight < session.LastBlockIndex && !Blockchain.Singleton.HeaderCacheFull)
             {
-                if (lastTaskIndex >= highestBlockIndex || lastTaskIndex >= Blockchain.Singleton.Height + InvPayload.MaxHashesCount) break;
-                if (!AssignSyncTask(++lastTaskIndex)) break;
+                session.Tasks[HeaderTaskHash] = DateTime.UtcNow;
+                IncrementGlobalTask(HeaderTaskHash);
+                session.RemoteNode.Tell(Message.Create(MessageCommand.GetHeaders, GetBlockByIndexPayload.Create(Blockchain.Singleton.HeaderHeight)));
             }
-        }
-
-        private void SendPingMessage()
-        {
-            TrimmedBlock block;
-            using (SnapshotCache snapshot = Blockchain.Singleton.GetSnapshot())
-            {
-                block = NativeContract.Ledger.GetTrimmedBlock(snapshot, NativeContract.Ledger.CurrentHash(snapshot));
-            }
-
-            foreach (KeyValuePair<IActorRef, TaskSession> item in sessions)
+            else if (Blockchain.Singleton.Height < session.LastBlockIndex)
             {
-                var node = item.Key;
-                var session = item.Value;
-
-                if (session.ExpireTime < TimeProvider.Current.UtcNow ||
-                     (block.Index >= session.LastBlockIndex &&
-                     TimeProvider.Current.UtcNow.ToTimestampMS() - PingCoolingOffPeriod >= block.Timestamp))
+                UInt256 hash = Blockchain.Singleton.CurrentBlockHash;
+                using (SnapshotCache snapshot = Blockchain.Singleton.GetSnapshot())
                 {
-                    if (session.InvTasks.Remove(MemPoolTaskHash))
+                    for (uint i = Blockchain.Singleton.Height + 1; i <= Blockchain.Singleton.HeaderHeight; i++)
                     {
-                        node.Tell(Message.Create(MessageCommand.Mempool));
+                        hash = Blockchain.Singleton.GetBlockHash(i, snapshot);
+                        if (!globalTasks.ContainsKey(hash))
+                        {
+                            hash = Blockchain.Singleton.GetBlockHash(i - 1, snapshot);
+                            break;
+                        }
                     }
```

Comment:
The `header_cache` may be changed while we are in the loop. It's not thread safe.

Reply:
fixed

Reply:
Adding lock on `IndexedQueue` is useless.

Reply:
I have reverted it. Please wait for me to add locks on other places.